File: plugins-community/ascendc-ops-lab-developer/workflows/templates/archive_tasks/matmul_leakyrelu/kernel/op_host/matmul_leakyrelu.cpp
/*
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "torch_kernel_helper.h"
#include "tiling/platform/platform_ascendc.h"
#include "aclrtlaunch_matmul_leakyrelu_fp16.h"
#include "aclrtlaunch_matmul_leakyrelu_int8.h"
#include "matmul_leakyrelu_tiling.h"

namespace ascend_kernel {

static inline int64_t GetWorkspaceSize(const MatmulLeakyReluTiling *tiling, int accElementSize,
                                       uint32_t usedCoreNum)
{
    return tiling->baseM * tiling->baseN * WORKSPACE_DEPTH * accElementSize * usedCoreNum;
}

at::Tensor matmul_leakyrelu(const at::Tensor &a, const at::Tensor &b)
{
    TORCH_CHECK(a.dim() == 2, "matmul_leakyrelu a must be two-dimensional.");
    TORCH_CHECK(b.dim() == 2, "matmul_leakyrelu b must be two-dimensional.");
    TORCH_CHECK(a.sizes()[1] == b.sizes()[0], "matmul_leakyrelu k must be same.");
    TORCH_CHECK(a.scalar_type() == b.scalar_type(), "matmul_leakyrelu a and b must have same dtype.");

    uint32_t usedCoreNum = 20;
    uint32_t m = static_cast<uint32_t>(a.sizes()[0]);
    uint32_t n = static_cast<uint32_t>(b.sizes()[1]);
    uint32_t k = static_cast<uint32_t>(a.sizes()[1]);

    at::Tensor c = at::empty({static_cast<int64_t>(m), static_cast<int64_t>(n)},
                             at::device(at::kPrivateUse1).dtype(at::kFloat));

    at::Tensor t = at::empty({static_cast<int64_t>(sizeof(MatmulLeakyReluTiling))},
                             at::device(at::kCPU).dtype(at::kByte));
    auto *tiling_ptr = reinterpret_cast<MatmulLeakyReluTiling *>(t.data_ptr());
    tiling_ptr->M = static_cast<int32_t>(m);
    tiling_ptr->N = static_cast<int32_t>(n);
    tiling_ptr->K = static_cast<int32_t>(k);
    tiling_ptr->baseM = DEFAULT_BASE_M;
    tiling_ptr->baseN = DEFAULT_BASE_N;
    tiling_ptr->baseK = DEFAULT_BASE_K;
    tiling_ptr->l1Prefetch = DEFAULT_L1_PREFETCH;
    auto tiling_npu = t.to(at::kPrivateUse1);

    uint32_t blockDim = usedCoreNum;

    if (a.scalar_type() == at::kHalf) {
        int accElementSize = sizeof(float);
        int64_t workSpaceSize = GetWorkspaceSize(tiling_ptr, accElementSize, usedCoreNum);
        at::Tensor w = at::empty({workSpaceSize}, at::device(at::kPrivateUse1).dtype(at::kByte));
        EXEC_KERNEL_CMD(matmul_leakyrelu_fp16, blockDim, a, b, c, w, tiling_npu);
    } else if (a.scalar_type() == at::kChar) {
        int accElementSize = sizeof(int32_t);
        int64_t workSpaceSize = GetWorkspaceSize(tiling_ptr, accElementSize, usedCoreNum);
        at::Tensor w = at::empty({workSpaceSize}, at::device(at::kPrivateUse1).dtype(at::kByte));
        EXEC_KERNEL_CMD(matmul_leakyrelu_int8, blockDim, a, b, c, w, tiling_npu);
    } else {
        TORCH_CHECK(false, "matmul_leakyrelu unsupported dtype, expected float16 or int8.");
    }

    return c;
}

} // namespace ascend_kernel

File: plugins-community/tilelang2ascendc-ops-generator/workflows/templates/archive_tasks/matmul_leakyrelu/design/tile_level/matmul_leakyrelu.py
#!/usr/bin/env python3
# ----------------------------------------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------------------------------------

from dataclasses import dataclass

import tilelang
import tilelang.language as T
from tilelang.intrinsics import make_zn_layout


pass_configs = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}


@dataclass
class MatmulConfig:
    m_size: int
    n_size: int
    k_size: int
    dtype: str = "float16"
    accum_dtype: str = "float"
    negative_slope: float = 0.01


@tilelang.jit(out_idx=[2], workspace_idx=3, pass_configs=pass_configs)
def matmul_leakyrelu(cfg):
    m_size, n_size, k_size, dtype, accum_dtype, negative_slope = (
        cfg.m_size, cfg.n_size, cfg.k_size, cfg.dtype, cfg.accum_dtype, cfg.negative_slope
    )
    base_m, base_n, base_k, l1_prefetch = 128, 128, 128, 4
    num_physical_cores = 20
    if m_size % base_m != 0:
        raise ValueError(f"M={m_size} must be divisible by baseM={base_m}")
    if n_size % base_n != 0:
        raise ValueError(f"N={n_size} must be divisible by baseN={base_n}")
    if k_size % (l1_prefetch * base_k) != 0:
        raise ValueError(f"K={k_size} must be divisible by l1Prefetch*baseK={l1_prefetch * base_k}")
    m_num = m_size // base_m
    n_num = n_size // base_n
    total_blocks = m_num * n_num
    if total_blocks <= 0:
        raise ValueError(f"total_blocks must be positive, got {total_blocks}")
    used_core_num = min(num_physical_cores, total_blocks)
    tasks_per_core = (total_blocks + used_core_num - 1) // used_core_num
    vec_num = 2
    use_float_accum = accum_dtype == "float"
    negative_slope_const = T.float32(negative_slope)

    @T.prim_func
    def main(  
        a_mat: T.Tensor((m_size, k_size), dtype),
        b_mat: T.Tensor((k_size, n_size), dtype),
        c_mat: T.Tensor((m_size, n_size), "float32"),
        # Persistent-kernel workspace: one tile buffer per physical core.
        workspace: T.Tensor((used_core_num, base_m, base_n), accum_dtype),
    ):
        with T.Kernel(used_core_num, is_npu=True) as (cid, vid):
            core_idx = cid

            a_l1 = T.alloc_L1((base_m, l1_prefetch * base_k), dtype)
            b_l1 = T.alloc_L1((l1_prefetch * base_k, base_n), dtype)
            T.annotate_layout({
                a_l1: make_zn_layout(a_l1),
                b_l1: make_zn_layout(b_l1),
            })
            a_l0 = T.alloc_L0A((base_m, base_k), dtype)
            b_l0 = T.alloc_L0B((base_k, base_n), dtype)
            c_l0 = T.alloc_L0C((base_m, base_n), accum_dtype)

            c_accum_ub = T.alloc_ub((base_m // vec_num, base_n), accum_dtype)
            c_out_ub = T.alloc_ub((base_m // vec_num, base_n), "float32")

            # Fixed physical cores; each core walks a contiguous task range.
            def _process_k_tile(m_idx, n_idx):
                loop_k = k_size // (l1_prefetch * base_k)
                for k in T.serial(loop_k):
                    outer = k * l1_prefetch
                    T.copy(a_mat[m_idx * base_m, outer * base_k], a_l1)
                    T.copy(b_mat[outer * base_k, n_idx * base_n], b_l1)
                    for kk in T.serial(l1_prefetch):
                        T.copy(a_l1[0, kk * base_k], a_l0)
                        T.copy(b_l1[kk * base_k, 0], b_l0)
                        T.mma(
                            a_l0,
                            b_l0,
                            c_l0,
                            init=T.And(k == 0, kk == 0),
                        )

            def _process_compute_tile(task_id, m_idx, n_idx):
                with T.Scope("C"):
                    if task_id < total_blocks:
                        _process_k_tile(m_idx, n_idx)
                        T.copy(c_l0, workspace[core_idx, 0, 0])
                        T.set_cross_flag("FIX", 0)

            def _process_vector_tile(task_id, m_idx, n_idx):
                with T.Scope("V"):
                    if task_id < total_blocks:
                        T.wait_cross_flag(0)
                        T.copy(
                            workspace[core_idx, vid * base_m // vec_num, 0],
                            c_accum_ub,
                        )
                        if use_float_accum:
                            T.copy(c_accum_ub, c_out_ub)
                        else:
                            T.tile.cast(
                                c_out_ub,
                                c_accum_ub,
                                mode="CAST_NONE",
                                count=(base_m // vec_num) * base_n,
                            )
                        T.tile.leaky_relu(c_out_ub, c_out_ub, negative_slope_const)
                        T.copy(c_out_ub, c_mat[m_idx * base_m + vid * base_m // vec_num, n_idx * base_n])

            def _process_tile(local_idx):
                task_id = core_idx * tasks_per_core + local_idx
                m_idx = task_id // n_num
                n_idx = task_id % n_num
                _process_compute_tile(task_id, m_idx, n_idx)
                _process_vector_tile(task_id, m_idx, n_idx)

            for local_idx in T.serial(tasks_per_core):
                _process_tile(local_idx)

    return main
